### PR TITLE
Fix flex attention captured vector gradient scatters

### LINKED ISSUE #148112: MLA with Learnable RoPE Tensors is Broken with Flex Attention
### 🐛 Describe the bug

I'm encountering an `AssertionError` during the backward pass when using `flex_attention` if the tensors used require gradients.

```python
import torch
from torch import Tensor
from torch.nn.attention.flex_attention import flex_attention

torch._inductor.config.unroll_reductions_threshold = 65

def generate_mla_rope_score_mod(query_rope: Tensor, key_rope: Tensor, num_heads: int, scale: float = 1.0):
    def mla_rope_score_mod(score: Tensor, b: Tensor, h: Tensor, q_idx: Tensor, kv_idx: Tensor) -> Tensor:
        return score + (scale * torch.dot(query_rope[b, h, q_idx], key_rope[b, h // num_heads, kv_idx]))

    return mla_rope_score_mod

def main(device: str = "cuda"):
    B, H, SEQ_LEN, LATENT_HEAD_DIM, ROPE_HEAD_DIM = 1, 128, 8, 512, 64

    query = torch.rand(B, H, SEQ_LEN, LATENT_HEAD_DIM, device=device, requires_grad=True)
    key = torch.rand(B, H, SEQ_LEN, LATENT_HEAD_DIM, device=device, requires_grad=True)
    value = torch.rand(B, H, SEQ_LEN, LATENT_HEAD_DIM, device=device, requires_grad=True)

    query_pe = torch.rand(B, H, SEQ_LEN, ROPE_HEAD_DIM, device=device, requires_grad=True)
    key_pe = torch.rand(B, H, SEQ_LEN, ROPE_HEAD_DIM, device=device, requires_grad=True)

    score_mod = generate_mla_rope_score_mod(query_pe, key_pe, ROPE_HEAD_DIM)
    out = torch.compile(flex_attention)(query, key, value, score_mod=score_mod)
    out.sum().backward()

main()
```

```
AssertionError:
  File "/home/ubuntu/.local/lib/python3.10/site-packages/torch/_inductor/lowering.py", line 3440, in fn
    assert len(idx) == len(output_size)
```

The full traceback shows that it's happening in the `flex_attention_backward` function within the inductor's kernel module.

I can confirm it works if you don't require gradients for `query_pe` and `key_pe`, but it seems like it should work if you do (and I have an adjacent application where they need to require gradients).

### Versions

'2.7.0a0+gitce805a5'

cc @chauhang @penguinwu @ydwu4 @bdhirsh @bobrenjc93 @aorenste @Chillee @drisspg @yanboliang @BoyuanFeng @liangel-02 @howardzhang-cv @zou3519

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185784

FlexAttention backward lowers gradients for tensors captured by score_mod via
zeros-and-scatter subgraphs. The Triton template path for these scatter
subgraphs passed no loop variables to Scatter.store_output, which only worked
for scalar captured gradients. A score_mod such as an MLA RoPE dot product
produces vector-valued captured gradients, so the scatter output indexer saw too
few indices and failed during lowering with an AssertionError.

Emit trailing scatter range indices for captured-gradient scatter subgraphs and
expand fixed template inputs/masks across those trailing dimensions. Non-power-
of-two trailing ranges are padded to Triton-friendly powers of two, with clamped
indices and masks so padded lanes do not write. CSE is invalidated at each
scatter-graph boundary because fixed-input expressions can have the same text
while carrying different trailing shape metadata for different captured tensors.

I considered keeping the fix local to FlexAttention's score_mod graph, but the
bug is in the template scatter lowering contract: Scatter.store_output needs the
range variables that describe the scatter payload. Fixing it there keeps scalar
captured gradients unchanged and handles vector captured gradients consistently.

Fixes #148112
Generated by my agent

Benchmark Results:
- Existing scalar captured-gradient path, baseline main loaded from
  HEAD:torch/_inductor/select_algorithm.py:
  `TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_148112_bench_baseline TRITON_CACHE_DIR=/tmp/triton_148112_bench_baseline timeout 900 python - <<'PY' ...B=1,H=4,S=64,D=64,float16 scale[0] score_mod, 3 warmup + 20 timed CUDA-event reps... PY`
  Result: `baseline_scalar_forward_backward_ms=0.5243`.
- Same benchmark with patched code:
  `TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_148112_bench_patched TRITON_CACHE_DIR=/tmp/triton_148112_bench_patched timeout 900 python - <<'PY' ...same scalar captured-gradient benchmark... PY`
  Result: `patched_scalar_forward_backward_ms=0.5020`.

Test Plan:
- Exact issue repro before fix failed with the reported AssertionError; after fix
  passed with `ok torch.Size([1, 128, 8, 512]) ...`.
- `TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_148112_vector_extents4 TRITON_CACHE_DIR=/tmp/triton_148112_vector_extents4 timeout 600 python test/inductor/test_flex_attention.py -k captured_vector_grads_different_trailing_extents -v`
- `TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_148112_learnable_rope3 TRITON_CACHE_DIR=/tmp/triton_148112_learnable_rope3 timeout 600 python test/inductor/test_flex_attention.py -k learnable_rope -v`
- `TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_148112_scalar2 TRITON_CACHE_DIR=/tmp/triton_148112_scalar2 timeout 600 python test/inductor/test_flex_attention.py -k captured_scalar_grad -v`
- `timeout 900 lintrunner -a`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased this onto current main and addressed an internal review cleanup by making the scatter IR contract explicit instead of falling back when range metadata is missing. Local lint and the focused FlexAttention regressions still pass; CI is rerunning on the updated branch.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
